File: src/main/java/dev/orion/users/application/usecases/CreateUserUC.java
/**
 * @License
 * Copyright 2024 Orion Services @ https://github.com/orion-services
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.orion.users.application.usecases;

import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.validator.routines.EmailValidator;

import dev.orion.users.application.interfaces.CreateUserUCI;
import dev.orion.users.enterprise.model.User;

public class CreateUserUC implements CreateUserUCI {

    /** The minimum size of the password required. */
    private static final int SIZE_PASSWORD = 8;

    /**
     * Creates a user in the service (UC: Create the user).
     *
     * @param name     : The name of the user
     * @param email    : The e-mail of the user
     * @param password : The password of the user
     * @return An User object
     */
    @Override
    public User createUser(final String name, final String email,
            final String password) {
        if (name.isBlank() || !EmailValidator.getInstance().isValid(email)
                || password.isBlank()) {
            throw new IllegalArgumentException(
                    "Blank arguments or invalid e-mail");
        } else {
            if (password.length() < SIZE_PASSWORD) {
                throw new IllegalArgumentException(
                        "Password less than eight characters");
            } else {
                // String secretKey = twoFactorAuthHandler.generateSecretKey();
                User user = new User();
                // user.setSecret2FA(secretKey);
                user.setName(name);
                user.setEmail(email);
                user.setPassword(encryptPassword(password));
                user.setEmailValid(false);
                return user;
            }
        }
    }

    /**
     * Creates a user in the service (UC: Authenticate With Google).
     *
     * @param name         : The name of the user
     * @param email        : The e-mail of the user
     * @param isEmailValid : Informs if the e-mail is valid
     * @return An User object
     */
    @Override
    public User createUser(final String name, final String email,
            final Boolean isEmailValid) {
        if (name.isBlank() || !EmailValidator.getInstance().isValid(email)) {
            throw new IllegalArgumentException(
                    "Blank arguments or invalid e-mail");
        } else {
            User user = new User();
            user.setName(name);
            user.setEmail(email);
            user.setEmailValid(isEmailValid);
            return user;
        }
    }

    /**
     * Encrypts the password with SHA-256.
     *
     * @param password : The password to be encrypted
     * @return The encrypted password
     */
    private String encryptPassword(final String password) {
        return DigestUtils.sha256Hex(password);
    }

}

File: src/main/kotlin/dev/orion/users/application/usecases/CreateUserUC.kt
/**
 * @License
 * Copyright 2025 Orion Services @ https://orion-services.dev
 */
package dev.orion.users.application.usecases

import dev.orion.users.application.port.`in`.CreateUserUCI
import dev.orion.users.application.utils.PasswordValidator
import dev.orion.users.domain.model.User
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import org.apache.commons.codec.digest.DigestUtils
import org.apache.commons.validator.routines.EmailValidator

@ApplicationScoped
open class CreateUserUC
    @Inject
    constructor() : CreateUserUCI {
        override fun createUser(
            name: String,
            email: String,
            password: String,
        ): User {
            if (name.isEmpty() || !EmailValidator.getInstance().isValid(email) || password.isEmpty()) {
                throw IllegalArgumentException("Blank arguments or invalid e-mail")
            }
            PasswordValidator.validatePasswordOrThrow(password)
            val user = User()
            user.name = name
            user.email = email
            user.password = encryptPassword(password)
            user.emailValid = false
            return user
        }

        override fun createUser(
            name: String,
            email: String,
            isEmailValid: Boolean,
        ): User {
            if (name.isBlank() || !EmailValidator.getInstance().isValid(email)) {
                throw IllegalArgumentException("Blank arguments or invalid e-mail")
            }
            val user = User()
            user.name = name
            user.email = email
            user.emailValid = isEmailValid
            return user
        }

        private fun encryptPassword(password: String): String = DigestUtils.sha256Hex(password)
    }
